cpp/src/arrow/ipc/writer.cc

```diff
@@ -263,6 +263,14 @@ class RecordBatchSerializer {
     out_->body_length = offset - buffer_start_offset_;
     DCHECK(bit_util::IsMultipleOf8(out_->body_length));
 
+    // copy given record batch's schema metadata to the serializer for serialization
+    auto const &metadata = batch.schema()->metadata();
+    if (metadata) {
+      for (int64_t i = 0; i < metadata->size(); ++i) {
+        AppendCustomMetadata(metadata->key(i), metadata->value(i));
+      }
+    }
```

Comment:
When writing a batch, its batch specific metadata will be copied into the serializer's internal metadata variable. Previously, we have the `AppendCustomMetadata` API, but it is never called, so the serializer's internal metadata variable is always empty.